### PR TITLE
Make EphemeralFile::write_blob function async

### PR DESCRIPTION
**Pull request by arpad-m**
## Problem

The `EphemeralFile::write_blob` function accesses the page cache internally. We want to require `async` for these accesses in #5023.

## Summary of changes

This removes the implementaiton of the `BlobWriter` trait for `EphemeralFile` and turns the `write_blob` function into an inherent function. We can then make it async as well as the `push_bytes` function. We move the `SER_BUFFER` thread-local into the `InMemoryLayerInner` so that the same buffer can be accessed by different threads as the async is (potentially) moved between threads.

Part of #4743, preparation for #5023.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by arpad-m** [comment]
Tbh this was easier than I thought. I thought it was more involved, that's why I made it a separate change.

File: pageserver/src/tenant/ephemeral_file.rs
Comment (koivunej) [inline pageserver/src/tenant/ephemeral_file.rs:93]: Should this be async as well or does it suffice that `write_blob` is async?

File: pageserver/src/tenant/storage_layer/inmemory_layer.rs
Comment (koivunej) [inline pageserver/src/tenant/storage_layer/inmemory_layer.rs:74]: Potentially risky change, because instead of fixed amount of buffers stuck to threads around we have `N` (=timelines). I don't think this could be [`task_local`](https://docs.rs/tokio/latest/tokio/macro.task_local.html) because we should have the same amount of those tasks.

Is the next to step make the `EphmeralFile::write_blob::Writer` into `AsyncWrite`? I don't think there has every been a need to buffer these, but making a TLS buffer is too convenient. Or does the `write_blob` need to be memcpy fast, or did it need to be for some synchronization issues? Would imagine not, the action is limited to single cell in the pagecache.

Comment (koivunej) [reply pageserver/src/tenant/storage_layer/inmemory_layer.rs:74]: To explain, there's double-buffering with the write_blob::Writer and this ser_buffer, it's unncessary.

> Or does the `write_blob` need to be memcpy fast

Of course it does not need to be because we are doing the small memcpies already to `ser_buffer`, so that we can do the fewer large memcpies to page cache buffer.

Comment (koivunej) [reply pageserver/src/tenant/storage_layer/inmemory_layer.rs:74]: 75MB per 10k timelines at 8k. We could probably afford this as is. Should we prealloc to 8kB or could we have larger..?

File: pageserver/src/tenant/ephemeral_file.rs
Comment (arpad-m) [reply pageserver/src/tenant/ephemeral_file.rs:93]: This is the actual function that has to be async, as it accesses the page cache internally. It's the same code that you [asked about in the other PR](https://github.com/neondatabase/neon/pull/5023#discussion_r1298842113).

File: pageserver/src/tenant/storage_layer/inmemory_layer.rs
Comment (arpad-m) [reply pageserver/src/tenant/storage_layer/inmemory_layer.rs:74]: If it's bounded by 8 KB we could also just have that on the stack instead, i.e. through SmallVec.

Comment (arpad-m) [reply pageserver/src/tenant/storage_layer/inmemory_layer.rs:74]: Mhh thinking of it, this seems like one could either use tokio's `task_local` macro (although I'm not sure if even that one supports access over await points), or one could just use `smallvec` with a reasonable default (like 128 bytes). If a value is above 128 bytes, this should happen rarely, we can afford an allocation.

It's important to note that anything stored on the stack by a task will be stored somewhere, so it contributes to memory overhead just as easily.

Comment (arpad-m) [reply pageserver/src/tenant/storage_layer/inmemory_layer.rs:74]: > Is the next to step make the `EphmeralFile::write_blob::Writer` into `AsyncWrite`?

No, this is not my plan. The `Writer` is just a struct. There is no trait involvement, and I would like to keep it that way. My plan is to make the page cache access in `push_bytes` async, and for that `push_bytes` needs to be async as well.

File: pageserver/src/tenant/ephemeral_file.rs
Comment (problame) [inline pageserver/src/tenant/ephemeral_file.rs:181]: Please avoid useless code movement.

File: pageserver/src/tenant/storage_layer/inmemory_layer.rs
Comment (problame) [reply pageserver/src/tenant/storage_layer/inmemory_layer.rs:74]: Hm, would be nice if the be_coder() could be an `AsyncRead`

Comment (arpad-m) [reply pageserver/src/tenant/storage_layer/inmemory_layer.rs]: Bincode 1.0 has no async serialization, and it looks like this is also not going to come for bincode 2: https://github.com/bincode-org/bincode/issues/519#issuecomment-1061925868

Comment (koivunej) [reply pageserver/src/tenant/storage_layer/inmemory_layer.rs]: The change is now to use SmallVec of 128 bytes. Did you test how large writes happen to these in our test suite? Well, I guess `pageserver::walrecord` the unpacked ones could fit into that nicely.

Comment (koivunej) [reply pageserver/src/tenant/storage_layer/inmemory_layer.rs]: I am good to resolve this.

**Review by koivunej — APPROVED** [review]

**Comment by arpad-m** [comment]
Will merge latest main once  #5068 is merged and then merge this PR.

File: pageserver/src/tenant/storage_layer/inmemory_layer.rs
Comment (problame) [inline pageserver/src/tenant/storage_layer/inmemory_layer.rs:273]: Is this 128 magic number chosen based on some deep insight or just shot from the hip? :) 

I guess we'd want to cover the largest `Value::WalRecord(NeonWalRecord)` variant?

Comment (arpad-m) [reply pageserver/src/tenant/storage_layer/inmemory_layer.rs:273]: I just took a guess at how large values typically are in databases, and I'd wager most should be below. Note that we care about the deep size here of the values, as we serialize them, so even though `Value::Image` only stores four pointers, its content can be way larger.

If you have any data about the distribution of the value sizes, that would be very useful.

Comment (arpad-m) [reply pageserver/src/tenant/storage_layer/inmemory_layer.rs:273]: I ran some tests, according to them a limit of 128 means we don't allocate in three quarters of cases. See [above](https://github.com/neondatabase/neon/pull/5056#discussion_r1301975061).

File: pageserver/src/tenant/storage_layer/inmemory_layer.rs
Comment (koivunej) [reply pageserver/src/tenant/storage_layer/inmemory_layer.rs]: The over 14k is interesting. Do you have idea what that was?

File: pageserver/src/tenant/storage_layer/inmemory_layer.rs
Comment (koivunej) [reply pageserver/src/tenant/storage_layer/inmemory_layer.rs]: > 3 bytes overhead
This is why I was thinking about the alternative ways to write and not copy everything to place A and then (in parts) to B.

**Review by problame — APPROVED** [review]
Good job with the analysis in https://github.com/neondatabase/neon/pull/5056#discussion_r1301975061

I don't know how representative the test suite is, would be interesting to also do a pgbench run. 

It would be nice to have a way to cheaply measure the `SmallVec::spilled` rate in a metric.

I guess putting a globally shared atomic on this hot path is counter the goal of reducing overhead, though 🤷

So, let's fly blind?

**Comment by arpad-m** [comment]
> I don't know how representative the test suite is, would be interesting to also do a pgbench run.

There are a bunch of pgbench invocations in the test suite:

```
$ rg -l pgbench test_runner/regress/
test_runner/regress/test_threshold_based_eviction.py
test_runner/regress/test_disk_usage_eviction.py
test_runner/regress/test_branching.py
test_runner/regress/test_pageserver_restarts_under_workload.py
test_runner/regress/test_duplicate_layers.py
test_runner/regress/test_compatibility.py
test_runner/regress/test_gc_cutoff.py
```

> I guess putting a globally shared atomic on this hot path is counter the goal of reducing overhead, though

Personally I think when we really copy over a value of like 256 bytes or more, then doing an allocation isn't as much overhead relatively to that copy.

**Comment by arpad-m** [comment]
CI has been failing five times already with the same error:

```
WARN[2023-08-22T22:51:04Z] Image 'neondatabase/compute-node-v15:5944521140' is not a file and couldn't be found in the container runtime 
ERRO[2023-08-22T22:51:04Z] Failed to import image(s) into cluster 'neon-local': no valid images specified 
WARN[2023-08-22T22:51:04Z] At least one error occured while trying to import the image(s) into the selected cluster(s) 
```

It doesn't sound like it was caused by my PR, but would love to be wrong.

**Comment by arpad-m** [comment]
Now, CI is giving errors like:

```
 #3 [internal] load metadata for docker.io/library/alpine:3.14
#3 ERROR: failed to authorize: failed to fetch anonymous token: Get "https://auth.ipv6.docker.com/token?scope=repository%3Alibrary%2Falpine%3Apull&service=registry.docker.io": dial tcp [2600:1f18:2148:bc01:c4:4967:28eb:1824]:443: connect: cannot assign requested address
```

This seems to also occur for CI runs of other currently open PRs.

**Pull request merged**
